### PR TITLE
Add __traits(isDynamicArray, T)

### PR DESCRIPTION
**Pull request by nordlow**
Discussion at https://forum.dlang.org/post/nhnofvzmijywsgqyyibw@forum.dlang.org

### CODE REVIEW

File: test/runnable/traits.d [hunk 1]
```diff
@@ -258,7 +258,45 @@ void test6()
 
 /********************************************************/
 
-void test7()
+void test7a()
+{
+    assert(__traits(isArray) == false);
```
Comment (BBasile) [inline test/runnable/traits.d]: you should get a compilation error for this. Argument count checking should lead to compilation error.

Comment (BBasile) [reply test/runnable/traits.d]: well only when there's a mismatch of course

**Comment by Geod24** [comment]
Why don't you use an `is` expression instead ?

File: test/runnable/traits.d (hunk 1 above)
Comment (nordlow) [reply test/runnable/traits.d]: All these tests were copied from the `__traits(isStaticArray, T)`-tests. Shouldn't we fix this problem in another PR?

**Comment by nordlow** [comment]
@Geod24, I've just copied the logic from `isStaticArray` to corresponding `isArray`. In what way should I use an `is`-expression?

File: test/runnable/traits.d (hunk 1 above)
Comment (BBasile) [reply test/runnable/traits.d]: in theory traits argument count is checked (there was an issue about this that i even opened IIRC) so maybe that this issue was not solved correctly.

**Comment by Geod24** [comment]
@nordlow : Instead of adding a traits to the compiler:
```d
template isArray (T)
{
    enum isArray = is(T == E[], E);
}

static assert(isArray!(char[]));
static assert(!isArray!(char[4]));
static assert(!isArray!(float));
```

**Comment by nordlow** [comment]
@edi33416 Are you saying that I should keep this as or modify the behaviour of `__traits(isArray)`?

File: test/runnable/traits.d [hunk 2]
```diff
@@ -1559,7 +1597,8 @@ int main()
     test4();
     test5();
     test6();
-    test7();
+    test7a();
+    test7b();
```
Comment (edi33416) [inline test/runnable/traits.d]: Is this ok? Maybe I'm missing something as I don't know much about the test system in dmd, but won't this recursively call `test7b`, generating a `stack limit exceeded` error?

File: test/runnable/traits.d [hunk 3]
```diff
@@ -258,7 +258,45 @@ void test6()
...
-void test7()
+void test7a()
+{
+    assert(__traits(isArray) == false);
+    assert(__traits(isArray, S) == false);
+    assert(__traits(isArray, C) == false);
+    assert(__traits(isArray, E) == false);
+    assert(__traits(isArray, void*) == false);
+    assert(__traits(isArray, void[]) == true);
+    assert(__traits(isArray, void[3]) == false);
```
Comment (edi33416) [inline test/runnable/traits.d]: I think that this should pass, as a static array is still an array from a user's perspective. The fact that it decays to something else (a `struct`) is an implementation detail.

**Review by edi33416 — CHANGES_REQUESTED** [review]

**Comment by Geod24** [comment]
@edi33416 : I just gave a quick template to match the implementation. I agree that static array should match, and you can do that with an is expression as well, just replacing `==` with `:`.

The main point is that there's already something to do this in the compiler, it's very accessible, easy to use, and quite well documented. So why should we add a trait ?

**Comment by edi33416** [comment]
> @edi33416 Are you saying that I should keep this as or modify the behaviour of `__traits(isArray)`? If so, can you describe in more detail how you think that the behaviour should change?

I strongly believe that `__traits(isArray, ...)` should behave exactly the same as the `Phobos` [isArray](https://github.com/dlang/phobos/blob/master/std/traits.d#L6618), as this should then be a drop-in replacement for the `std.traits.isArray` implementation.

If it won't work exactly as the `std.traits.isArray`, then it will just cause confusion amongst the users, which, in my opinion, is something that we don't want.

As a way to test your PR, I'd advise testing the changes by modifying the `std.traits.isArray` to be your `__traits(isArray, T)` and running the Phobos unittests.

**Comment by BBasile** [comment]
Yes, i was also expecting a replacement (alias this stuff like in std.traits handled or not ?).
So maybe failed attempt and close.

**Comment by edi33416** [comment]
> @edi33416 : I just gave a quick template to match the implementation. I agree that static array should match, and you can do that with an is expression as well, just replacing `==` with `:`.
> 
> The main point is that there's already something to do this in the compiler, it's very accessible, easy to use, and quite well documented. So why should we add a trait ?

Totally agree with you. IIRC I tried to change the Phobos implementation to `is(T : U[], U)`, but some tests were failing. I believe it had something to do with how `isDynamicArray` is implemented (it also checks for `alias this`; see `AliasThisTypeOf`)

**Comment by edi33416** [comment]
> The main point is that there's already something to do this in the compiler, it's very accessible, easy to use, and quite well documented. So why should we add a trait ?

The reason to add this as a trait is to have a complete set. We already have `isStaticArray`, which can be written as `is(T == U[n], U, size_t n)`.

I think that because of this, the users expect to also find an `isArray` and an `isDynamicArray`. I know I did.

**Comment by nordlow** [comment]
Are you saying we should add both `__traits(isDynamicArray)` and `__traits(isArray)`?

**Comment by edi33416** [comment]
> Are you saying we should add both `__traits(isDynamicArray)` and `__traits(isArray)`?

As separate PRs, but yes. That's my point of view, at least.
Let's see what the others think.

**Comment by BBasile** [comment]
> How should these behave when T being a aggregate type that behave as an array via alias this?

Sorry this is where i opt out. I thought that the idea was to handle that in the new trait.

**Comment by nordlow** [comment]
@BBasile are saying we shouldn't care about `alias this` in the these new `__traits(..., T)`?

**Comment by BBasile** [comment]
No i'm saying that i'm not sure about this change at all. i expected something else. Maybe because of a misunderstanding of me when i read your NG post.

File: test/runnable/traits.d (hunk 2 above)
Comment (RazvanN7) [reply test/runnable/traits.d]: Why would it call it recursively? Just expand the code a bit... this is in main().

Comment (edi33416) [reply test/runnable/traits.d]: Whops :laughing:

**Comment by nordlow** [comment]
Changed `__traits(isArray, T)` to be `true` also when `T` is a non-static array (slice).

**Comment by RazvanN7** [comment]
I think that the best course of action would be to add `isDynamicArray`. We already have `isStaticArray` and `isAssociativeArray`. Adding `isArray` is just going to cause confusion over what type of array are we checking.

Regarding aggregate declarations that have an aliased this member that is an array, IMHO the check should fail. Alias this means that the type is convertible to the alias member, not that it behaves exactly like that member. Moreover, `isStaticArray` yields false in the following code:

```d
struct B
{
	int[3] a;
    alias a this;
}

void main()
{
    B a;
 	import std.stdio : writeln;
    writeln(__traits(isStaticArray, a));
}
```

**Comment by nordlow** [comment]
@RazvanN7 ok with me. Shall I update the PR?

**Comment by RazvanN7** [comment]
Yes. Also, if this is going to get in it will require a changelog entry and a spec update

**Comment by nordlow** [comment]
Is the changelog in repo dmd and the spec update in repo dlang.org?

**Comment by RazvanN7** [comment]
There's a changelog directory in the dmd repo (in the root directory). You will find over there examples on how the changelog entry should look like.

For the spec, just go here: https://dlang.org/spec/traits.html and click on the improve this page button (upper right corner). No need to  clone the repo, it will be faster that way.

**Comment by nordlow** [comment]
Add spec update at https://github.com/dlang/dlang.org/pull/2516.

**Review by Geod24 — CHANGES_REQUESTED** [review]
I really don't see any advantage in that trait over `is(T == E[], E)` and think this should be closed.

**Comment by RazvanN7** [comment]
> I really don't see any advantage in that trait over is(T == E[], E) and think this should be closed.

That is also true for static arrays and associative, still we do have a trait for them. How do you explain that?

**Comment by Geod24** [comment]
> How do you explain that?

If you want to convince me that the reasoning that applied to `isStaticArray` and `isAssociativeArray` holds for `isDynamicArray`, you should bring that reasoning forward, not the other way around.

But I'll do your homework nonetheless.

Those traits were part of the first traits, [introduced 10 years ago](https://github.com/dlang/dmd/commit/e56081448fff1af90b8eda765d649a131bb19c3c#diff-0944663612b319fada07a9e3d4d07910R226), at the very early stage of D2.
Since it is before both our time, I would have a hard time explaining it, but it predates `is` expressions support for `is(T == E[], E)`. See `parse.c` at the same commit: https://github.com/dlang/dmd/blob/e56081448fff1af90b8eda765d649a131bb19c3c/src/parse.c#L4041-L4086
In D1, there was support for `is(T == char[])` (or any other array type), but AFAIK there was no way to match any array type.
Why `isDynamicArray` was not added at that time, I do not know, maybe it was forgotten and ended up being added directly to `IsExpression` because the need for type deconstruction arose.

**Comment by RazvanN7** [comment]
> you should bring that reasoning forward, not the other way around.

Maybe it wasn't obvious from how I phrased it, but that was the reasoning that I was trying to bring forward.

Anyway, you are indeed right. I just had a chat with Andrei and he is against this addition on the grounds that the `__traits` idiom was intentionally conceived to be ugly from a syntactic point of view to discourage the wide use for it. `std.traits` should be everyone's go to when it comes to traits.

So it seems that is is not going to get in. Sorry for your wasted effort @nordlow .

**Comment by nordlow** [comment]
My motivation behind this addition is to reduce the number of template instantations for the std.traits that can make use of this built-in trait.

**Comment by RazvanN7** [comment]
@nordlow Sorry, but there isn't any support from any of the core dmd members for this. I will close it for now, but if you manage to convince Andrei or Walter about this addition, you can always reopen the PR.

**Comment by nordlow** [comment]
There has been much talk about reducing number of template instantiations lately. This will further reduce the need for that when it comes to `std.traits.isDynamicArray` which is very popular trait. Is there still no interest in adding this builtin? @WalterBright, @andralex, @UplinkCoder?

**Comment by MoonlightSentinel** [comment]
> I really don't see any advantage in that trait over `is(T == E[], E)` and think this should be closed.

I second this.

Changed the implementation of `isDynamicArray` in dlang/phobos#7574

**Comment by andralex** [comment]
In fairness there's not much consistency in the traits. `isStaticArray` is there so one would expect things like `isDynamicArray` and `isPointer`. `isStaticArray` is also unnecessary because it's replaceable with `is(T == U[n], U, size_t n)`.

As long as the pattern matching is not considerably slower, I think adding this trait won't go a long way toward simplifying matters.

**Comment by andralex** [comment]
I approved https://github.com/dlang/phobos/pull/7574 which seems an obvious step forward in any case. @nordlow does this help your case?

**Comment by nordlow** [comment]
> I approved [dlang/phobos#7574](https://github.com/dlang/phobos/pull/7574) which seems an obvious step forward in any case. @nordlow does this help your case?

Let's have this Phobos-change merged and then I can rebenchmark against that change. Ok?

**Comment by UplinkCoder** [comment]
I will take a look tomorrow morning.

In general I would say that the complexity introduced by this PR is very manageable.
Especially since we already have an `isStaticArray` trait.

**Comment by nordlow** [comment]
> I will take a look tomorrow morning.
> 
> In general I would say that the complexity introduced by this PR is very manageable.
> Especially since we already have an `isStaticArray` trait.

Yes and this will still save cycles compared to the proposed Phobos update of `std.traits.isDynamicArray` this builtin trait is a no-op for the compiler.

**Comment by MoonlightSentinel** [comment]
> Yes and this will still save cycles compared to the proposed Phobos update of `std.traits.isDynamicArray`. Choosing the builtin trait will be a single cast away for the compiler.

Sure but the addition implies a slight penalty for many other `__traits(...)` expressions. Looking forward to some benchmarks.

**Comment by nordlow** [comment]
> Sure but the addition implies a slight penalty for many other `__traits(...)` expressions. Looking forward to some benchmarks.

I presume you are talking about the phobos addition.

Do you want me to benchmark before or after the phobos PR has been merged?

**Comment by MoonlightSentinel** [comment]
> I presume you are talking about the phobos addition.

I was refering to `traits(__isDynamicArray)`.
 
> Do you want me to benchmark before or after the phobos PR has been merged?

After. It should be compared to the `is(...)` approach.

**Comment by nordlow** [comment]
What was the motivation behind closing this? Shouldn't a benchmark be performed first? @MoonlightSentinel?

**Comment by MoonlightSentinel** [comment]
This trait is still redundant to other language facilities and doesn't bring value on its own*. The only argument in favor of this PR could be a potential performance gain from minimal changes. So a benchmark would definitly help your case here.

*i'm explicitly disregarding the `isStaticArray` argument because `is(...)` wasn't capable enough back then AFAICT.

**Comment by nordlow** [comment]
> This trait is still redundant to other language facilities and doesn't bring value on its own*. The only argument in favor of this PR could be a potential performance gain from minimal changes. So a benchmark would definitly help your case here.
> 
> *i'm explicitly disregarding the `isStaticArray` argument because `is(...)` wasn't capable enough back then AFAICT.

So which implementation of `isDynamicArray` do you want me to benchmark against?

**Comment by MoonlightSentinel** [comment]
Let's compare:
- `traits(isDynamicArray)` from this PR
- `isDynamicArray` from dlang/phobos#7574

This ignores the `enum` issue discussed in https://github.com/dlang/phobos/pull/7574#discussion_r467514715 because `traits(isDynamicArray)` and `is(T == U[], U)` must yield the same result (whatever that may be*).

* I still think that the current behaviour of `isDynamicArray` (and your PR) is wrong because they disregard that an `enum` is a distinct type - but maybe there is a good reason to do so

**Comment by nordlow** [comment]
>     * `isDynamicArray` from [dlang/phobos#7574](https://github.com/dlang/phobos/pull/7574)

Do you mean

```D
enum bool isDynamicArray(T) = is(T == U[], U);
```

or

```D
template isDynamicArray(T)
{
    static if (is(T == U[], U))
        enum bool isDynamicArray = true;
    else static if (is(T U == enum))
        enum bool isDynamicArray = isDynamicArray!U;
    else
        enum bool isDynamicArray = false;
}
```

?

**Comment by MoonlightSentinel** [comment]
The first:
```
enum bool isDynamicArray(T) = is(T == U[], U);
```

Benchmarking against the second implementation would be pretty pointless because it suffers from`static if` and recursive template instantiations.

**Comment by nordlow** [comment]
> Benchmarking against the second implementation would be pretty pointless because it suffers from`static if` and recursive template instantiations.

Got it.

**Comment by nordlow** [comment]
> Let's compare:
> 
>     * `traits(isDynamicArray)` from this PR
> 
>     * `isDynamicArray` from [dlang/phobos#7574](https://github.com/dlang/phobos/pull/7574)
> 
> 
> This ignores the `enum` issue discussed in [dlang/phobos#7574 (comment)](https://github.com/dlang/phobos/pull/7574#discussion_r467514715) because `traits(isDynamicArray)` and `is(T == U[], U)` must yield the same result - whatever that may be.
> 
> (I still think that the current behaviour of `isDynamicArray` (and your PR) is wrong because they disregard that an `enum` is a distinct type - but maybe there is a good reason to do so)

I guess we need to pull in some other folks for comments about whether enum of strings should be considered a `isDynamicArray` or not. I agree that an enum should not be considered a dynamic array, whereas its individual enumerators should be considered constants of underlying type, in your care `string` being a dynamic array.

**Comment by nordlow** [comment]
The builtin in trait is slightly faster:

> Benchmarking against the second implementation would be pretty pointless because it suffers from`static if` and recursive template instantiations.

```D
import std.meta : AliasSeq;

private static alias ScalarTypes = AliasSeq!(bool,
                                             char, wchar, dchar,
                                             byte, ubyte,
                                             short, ushort,
                                             int, uint,
                                             long, ulong,
                                             float, double, real,
                                             cfloat, cdouble, creal,
                                             ifloat, idouble, ireal);

private static enum qualifiers = AliasSeq!("", "const", "inout", "immutable");

static foreach (T; ScalarTypes)
{
    static foreach (U; ScalarTypes)
    {
        static foreach (V; ScalarTypes)
        {
            mixin("struct ",
                  T, "_" ,U, "_" ,V,
                  " {",
                  T, " t; ",
                  U, " u; ",
                  V, " v; ",
                  "}");
            static foreach (qualifier; qualifiers)
            {
                // static assert(__traits(isDynamicArray, mixin(qualifier, "(", T, "_", U, "_", V, ")")[]));
                // min over 10 runs: 1.38 s
                // static assert(is(mixin(qualifier, "(", T, "_", U, "_", V, ")")[] == X[], X)); // min over 10 runs: 1.42 s
            }
        }

    }
}
```

I meaured a minimum

- 1.38 s for __traits(isDynamicArray) vs
- 1.42 s for is(...)

over 15 runs.

**Comment by nordlow** [comment]
Ping.

**Comment by andralex** [comment]
Yes, `enum` types and `enum` values are different beasts.

We're looking at a 5% improvement of the test alone. The impact of the dynamic array test over a compilation's bottom line is small in the best of cases - maybe under 1%. Therefore, the impact of this PR on the bottom line is under 5% of 1%, i.e. 5 in 10,000. I doubt it's ever measurable outside specialized microbenchmarks.

I appreciate the effort, but if speed is a factor, a cold-hearted calculation suggests closing this PR and taking the experience accumulated to a more impactful endeavor.

**Comment by UplinkCoder** [comment]
> We're looking at a 5% improvement of the test alone. The impact of the dynamic array test over a compilation's bottom line is small in the best of cases - maybe under 1%. Therefore, the impact of this PR on the bottom line is under 5% of 1%, i.e. 5 in 10,000. I doubt it's ever measurable outside specialized microbenchmarks.

I see in `maybe` in that explantation. Which shows that you have not messured.
In the recent past improvments the memory usage of dmd have shown to have stayed under 1% improvement and yet they were accepted, altough they intoduced a new system and had a much wider bug attack surface than this PR.

`is`-expressions are generally more complex in the compiler logic than `__traits` are.
Therefore it would be welcome to move more capablities into traits.

@nordlow I'd suggest reopening.

**Comment by nordlow** [comment]
> @nordlow I'd suggest reopening.

How do I reopen?

**Comment by nordlow** [comment]
> > We're looking at a 5% improvement of the test alone. The impact of the dynamic array test over a compilation's bottom line is small in the best of cases - maybe under 1%. Therefore, the impact of this PR on the bottom line is under 5% of 1%, i.e. 5 in 10,000. I doubt it's ever measurable outside specialized microbenchmarks.
> 
> I see `maybe` in that explantation. Which hints of some doubts.
> 
> `is`-expressions are generally more complex in the compiler logic than `__traits` are.
> Therefore it would be welcome to move more capabilities into traits.
> 
> @nordlow I'd suggest reopening.

I forgot to mention that if we want to preserve the behaviour of `enum` support of `std.traits.isDynamicArray`, also currently supported by the implementation of `__traits(isDynamicArray, ...)` in this PR, we have to use the templated version

```D
template isDynamicArray(T)
{
    static if (is(T == U[], U))
        enum bool isDynamicArray = true;
    else static if (is(T U == enum))
        enum bool isDynamicArray = isDynamicArray!U;
    else
        enum bool isDynamicArray = false;
}
```

then the execution time jumps up to 3.19s for the non-builtin case as in

```D
private static alias ScalarTypes = AliasSeq!(bool,
                                             char, wchar, dchar,
                                             byte, ubyte,
                                             short, ushort,
                                             int, uint,
                                             long, ulong,
                                             float, double, real,
                                             cfloat, cdouble, creal,
                                             ifloat, idouble, ireal,
                                             string, wstring, dstring);

private static enum qualifiers = AliasSeq!("", "const", "inout", "immutable", "shared", "shared const");

static private template isDynamicArray(T)
{
    static if (is(T == U[], U))
        enum bool isDynamicArray = true;
    else static if (is(T U == enum))
        enum bool isDynamicArray = isDynamicArray!U;
    else
        enum bool isDynamicArray = false;
}

static foreach (T; ScalarTypes)
{
    static foreach (U; ScalarTypes)
    {
        static foreach (V; ScalarTypes)
        {
            mixin("struct ",
                  T, "_" ,U, "_" ,V,
                  " {",
                  T, " t; ",
                  U, " u; ",
                  V, " v; ",
                  "}");
            static foreach (qualifier; qualifiers)
            {
                version(useBuiltin)
                    static assert(__traits(isDynamicArray, mixin(qualifier, "(", T, "_", U, "_", V, ")")[])); // min over 10 runs: 2.62s.
                else
                    // static assert(is(mixin(qualifier, "(", T, "_", U, "_", V, ")")[] == X[], X)); // min over 10 runs: 2.75s
                    static assert(isDynamicArray!(mixin(qualifier, "(", T, "_", U, "_", V, ")")[])); // min over 10 runs: 3.19s
            }
        }
    }
}
```

**Comment by UplinkCoder** [comment]
@nordlow this is a nice benchmark and we can clearly see the template overhead.

You should be able to reopen by just clicking reopen at the bottom of this commnet box,
If you don't see that then it could be that you didn't close the PR.

**Comment by nordlow** [comment]
Further note that the overhead will be even greater for the non-built-in case for enum constants of dynamic arrays which will require two template instantiations per type.

**Comment by nordlow** [comment]
I can't find any reopen button anywhere. I guess I'm missing some permissions for this. I guess I have to create a new PR. @UplinkCoder?

**Comment by UplinkCoder** [comment]
@nordlow if you closed the PR, I think so. Yes you'd have to open a new PR.

**Comment by nordlow** [comment]
> @nordlow if you closed the PR, I think so. Yes you'd have to open a new PR.

Done.

**Comment by MoonlightSentinel** [comment]
> I forgot to mention that if we want to preserve the behaviour of enum support of std.traits.isDynamicArray, also currently supported by the implementation of __traits(isDynamicArray, ...) in this PR, we have to use the templated version

No, if the current implementation of `isDynamicArray` is correct we need to modify `is`. Having both yielding different results is hazardous.

**Comment by andralex** [comment]
> > We're looking at a 5% improvement of the test alone. The impact of the dynamic array test over a compilation's bottom line is small in the best of cases - maybe under 1%. Therefore, the impact of this PR on the bottom line is under 5% of 1%, i.e. 5 in 10,000. I doubt it's ever measurable outside specialized microbenchmarks.
> 
> I see `maybe` in that explantation. Which hints of some doubts.

Was qualifying my estimate, as anyone should. But the burden is not on me. If anything, the estimate was too generous. It is very unlikely to find an application that tests for dynamic arrays 1% of its build time. That would mean a combination of frequent calls to `isDynamicArray` and long absolute times for each use. Neither of these was demonstrated. What I can tell is there are 80 uses in all of Phobos' 300KLOC, including unittests, 

Per the latest microbenchmark, we're looking at 3.19s vs. 2.75s, i.e. 16%. That's still a far cry from indicating any impact. Why do we bother measuring if then we ignore the results?

> `is`-expressions are generally more complex in the compiler logic than `__traits` are.
> Therefore it would be welcome to move more capabilities into traits.

I disagree with this argument. By it we should take most of the work the programmer can do and hardcode it as compiler magic. This is exactly the opposite of what we should be doing. `__traits` has already become a dumping ground of punctual features with no rhyme or reason. We should be weary of adding to that.

**Comment by UplinkCoder** [comment]
> We should be weary of adding to that.
It is good to be weary of adding introspection features, since the stability of which can be hard to guarantee.

> Per the latest microbenchmark, we're looking at 3.19s vs. 2.75s, i.e. 16%. That's still a far cry from indicating any impact.
So you are saying that a 16% impact (best case arguably) is no impact?

**Comment by UplinkCoder** [comment]
This is starting to get fiery again which is not good

@andralex I would really like to understand your reasoning in depth.

Please do expand on how
`is(T == U[], U)`
is more readable than 
`__traits(isDynmaicArray, T)`

**Comment by andralex** [comment]
> Please do expand on how
> `is(T == U[], U)`
> is more readable than
> `__traits(isDynmaicArray, T)`

Let's see. This seems to be a false choice for two reasons. First, the real interface is `isDynamicArray!T` - the others are used as backend. Second, the two features are not identical; for example the first also offers the array element type. So the trait does not supplant `is` for arrays.

To be clear, I do agree with the consistency take: there is a `isStaticArray` trait, so a consistency/completeness/expectation argument can be built here. I also have a lot of respect for the work done here, it's very specialized and offered for free. But if the case jumps from performance to readability to whatever it looks like merging the feature is the foregone conclusion we build arguments around. Things should flow the other way.

**Comment by UplinkCoder** [comment]
@andralex 
I merely asked which one of the two expressions presented conveys the intent better, in your opinion.

**Pull request closed without merging**